src/main/scala/nodes/learning/KernelGenerator.scala

```diff
@@ -0,0 +1,163 @@
+package nodes.learning
+
+import scala.reflect.ClassTag
+
+import breeze.linalg._
+import breeze.math._
+import breeze.numerics._
+
+import org.apache.spark.rdd.RDD
+
+import pipelines.Logging
+import workflow.Transformer
+import utils._
+
+/**
+ * Base trait for functions that compute a kernel on a given dataset.
+ */
+trait KernelGenerator[T] {
+
+  /**
+   * Create a kernel transformer using the given dataset as one of the arguments
+   * to the kernel function. That is, if the kernel function is \phi(x, y), this
+   * binds one of the arguments using the given data.
+   *
+   * @param trainData training data matrix to be used for kernel transformation
+   * @return a kernel transformer that can be applied to train or test data
+   */
+  def fit(trainData: RDD[T]): KernelTransformer[T]
```

Comment:
The reason we can't do `KernelTransformer extends Transformer[T, DenseVector[Double]` is because we can't return a `RDD[DenseVector[Double]]` as an output from `apply(RDD[T])`. This is a limitation of our Transformer API which enforces the pattern to be row-oriented while the `KernelMatrix` here is column block oriented.  Or to put it another way the trait definition where `T => U` and `RDD[T] => RDD[U]` means that we can't use custom datastructures like KernelMatrix in the type definition.